src/include/duckdb/storage/compression/alprd/algorithm/alprd.hpp

```diff
@@ -0,0 +1,221 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/storage/compression/alprd/algorithm/alprd.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/storage/compression/patas/patas.hpp"
+#include "duckdb/storage/compression/alprd/alprd_constants.hpp"
+
+#include "duckdb/common/bitpacking.hpp"
+#include "duckdb/common/common.hpp"
+#include "duckdb/common/pair.hpp"
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/limits.hpp"
+
+namespace duckdb {
+
+namespace alp {
+
+template <class T, bool EMPTY>
+class AlpRDCompressionState {
+public:
+	using EXACT_TYPE = typename FloatingToExact<T>::type;
+
+	AlpRDCompressionState() : right_bw(0), left_bw(0), exceptions_count(0) {
+	}
+
+	void Reset() {
+		left_bp_size = 0;
+		right_bp_size = 0;
+		exceptions_count = 0;
+	}
+
+public:
+	uint8_t right_bw;
+	uint8_t left_bw;
+	uint16_t exceptions_count;
+	uint8_t right_parts_encoded[AlpRDConstants::ALP_VECTOR_SIZE * 8];
+	uint8_t left_parts_encoded[AlpRDConstants::ALP_VECTOR_SIZE * 8];
+	uint16_t left_parts_dict[AlpRDConstants::DICTIONARY_SIZE];
+	uint16_t exceptions[AlpRDConstants::ALP_VECTOR_SIZE];
+	uint16_t exceptions_positions[AlpRDConstants::ALP_VECTOR_SIZE];
+	idx_t left_bp_size;
+	idx_t right_bp_size;
+	unordered_map<uint16_t, uint16_t> left_parts_dict_map;
+};
+
+template <class T, bool EMPTY>
+struct AlpRDCompression {
+	using State = AlpRDCompressionState<T, EMPTY>;
+	using EXACT_TYPE = typename FloatingToExact<T>::type;
+	static constexpr uint8_t EXACT_TYPE_BITSIZE = sizeof(EXACT_TYPE) * 8;
+
+	/*
+	 * Estimate the bits per value of ALPRD within a sample
+	 */
+	static double EstimateCompressionSize(uint8_t right_bw, uint8_t left_bw, uint16_t exceptions_count,
+	                                      uint64_t sample_count) {
+		double exceptions_size =
+		    exceptions_count * ((AlpRDConstants::EXCEPTION_POSITION_SIZE + AlpRDConstants::EXCEPTION_SIZE) * 8);
+		double estimated_size = right_bw + left_bw + (exceptions_size / sample_count);
+		return estimated_size;
+	}
+
+	static double BuildLeftPartsDictionary(vector<EXACT_TYPE> values, uint8_t right_bw, uint8_t left_bw,
```

Comment:
`uint8_t left_bw` seems to be completely unused here?

Reply:
This was an error in the implementation from my side. Here I wrongly used always a dictionary of the same size. This is not optimised since there can be smaller bit widths if the dictionary is small enough. 

It is now properly implemented.